clang/tools/clang-linker-wrapper/ClangLinkerWrapper.cpp

```diff
@@ -633,37 +635,140 @@ static Expected<StringRef> runLLVMToSPIRVTranslation(StringRef InputTable,
   return *Output;
 }
 
-// Run clang-offload-wrapper
-static Expected<StringRef> runWrapper(StringRef &InputFile,
-                                      const ArgList &Args) {
-  // Create a new file to write the wrapped file to.
-  auto TempFileOrErr =
-      createOutputFile(sys::path::filename(ExecutableName), "bc");
-  if (!TempFileOrErr)
-    return TempFileOrErr.takeError();
-  Expected<std::string> ClangOffloadWrapperPath = findProgram(
-      "clang-offload-wrapper", {getMainExecutable("clang-offload-wrapper")});
-  if (!ClangOffloadWrapperPath)
-    return ClangOffloadWrapperPath.takeError();
+Expected<std::vector<char>> readBinaryFile(StringRef File) {
+  auto MBOrErr = MemoryBuffer::getFile(File, /*IsText*/ false,
+                                       /*RequiresNullTerminator */ false);
+  if (!MBOrErr)
+    return createFileError(File, MBOrErr.getError());
 
-  BumpPtrAllocator Alloc;
-  StringSaver Saver(Alloc);
+  auto &MB = *MBOrErr;
+  return std::vector<char>(MB->getBufferStart(), MB->getBufferEnd());
+}
 
-  SmallVector<StringRef, 8> CmdArgs;
-  CmdArgs.push_back(*ClangOffloadWrapperPath);
-  CmdArgs.push_back(Saver.save("-o=" + *TempFileOrErr));
-  llvm::Triple HostTriple(
+Expected<std::string> readTextFile(StringRef File) {
+  auto MBOrErr = MemoryBuffer::getFile(File, /*IsText*/ true,
+                                       /*RequiresNullTerminator */ true);
+  if (!MBOrErr)
+    return createFileError(File, MBOrErr.getError());
+
+  auto &MB = *MBOrErr;
+  return std::string(MB->getBufferStart(), MB->getBufferEnd());
+}
+
+Expected<std::unique_ptr<util::PropertySetRegistry>>
+readPropertyRegistryFromFile(StringRef File) {
+  auto MBOrErr = MemoryBuffer::getFile(File, /*IsText*/ true);
+  if (!MBOrErr)
+    return createFileError(File, MBOrErr.getError());
+
+  auto &MB = *MBOrErr;
+  return util::PropertySetRegistry::read(&*MB);
+}
+
+// The table format is the following:
+// [Code|Properties|Symbols]
+// a_0.bin|a_0.prop|a_0.sym
+// .
+// a_n.bin|a_n.prop|a_n.sym
+//
+// .bin extension might be a bc, spv or other native extension.
+Expected<SmallVector<SYCLImage>> readSYCLImagesFromTable(StringRef TableFile,
```

Comment:
In a subsequent change, we can try to create SYCLImage inside the sycl-post-link and pass it along. This will require us to replace sycl-post-link and SPIR-V Translator to be replaced with library calls.